bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -263,45 +266,63 @@ boolean doRefresh() {
-      // reached its configured timeout limit.
+  protected HeaderCacheElement refreshCredentialsWithRetry() {
+    BackOff backoff = null;
+    while (true) {
+      // Performs a return when:
+      //  1) An OAuth request was completed.  If the value is null, return an exception.
+      //  2) A non-IOException Exception is thrown - return an error status
+      //  3) All retries have been exhausted, i.e. when the Backoff.nextBackOffMillis() returns
+      //     BackOff.STOP
+      //  4) An interrupt occurs.
       try {
-        RetryState retryState = getRetryState(backoff);
-        if (retryState == RetryState.PerformRetry) {
-          return refreshCredentialsWithRetry(backoff);
+        infoLog("Refreshing the OAuth token");
+        AccessToken newToken = credentials.refreshAccessToken();
+        if (newToken == null) {
+          // The current implementations of refreshAccessToken() throws an IOException or return
+          // a valid token. This handling is future proofing just in case credentials returns null for
+          // some reason. This case was caught by a poorly coded unit test.
+          infoLog("Refreshed the OAuth token");
+          return new HeaderCacheElement(new IOException("Could not load the token for credentials: "
+              + credentials));
         } else {
+          // Success!
+          return new HeaderCacheElement(newToken);
+        }
+      } catch (IOException exception) {
+        warnLog("Got an unexpected IOException when refreshing google credentials.", exception);
+        // An IOException occurred. Retry with backoff.
+        if (backoff == null) {
+          // initialize backoff.
+          backoff = retryOptions.createBackoff();
+        }
+        // Given the backoff, either sleep for a short duration, or terminate if the backoff has
+        // reached its configured timeout limit.
+        try {
+          RetryState retryState = getRetryState(backoff);
+          if (retryState != RetryState.PerformRetry) {
+            return new HeaderCacheElement(exception);
+          } // else Retry.
+        } catch (IOException e) {
+          warnLog("Got an exception while trying to run backoff.nextBackOffMillis()", e);
           return new HeaderCacheElement(exception);
         }
-      } catch (IOException e) {
-        LOG.warn("Got an exception while trying to run backoff.nextBackOffMillis()", e);
-        return new HeaderCacheElement(exception);
+      } catch (Exception e) {
+        warnLog("Got an unexpected exception while trying to refresh google credentials.", e);
+        return new HeaderCacheElement(new IOException("Could not read headers", e));
       }
-    } catch (Exception e) {
-      LOG.warn("Got an unexpected exception while trying to refresh google credentials.", e);
-      return new HeaderCacheElement(new IOException("Could not read headers", e));
+    }
+  }
+
+  private void infoLog(String message) {
```

Comment:
I am fine with this, would be nice if we have something like this in "bigtable/config/Logger.java".